src/EventStore.Core/Services/RequestManager/RequestManagementService.cs

```diff
@@ -24,149 +25,182 @@ public class RequestManagementService : IHandle<SystemMessage.SystemInit>,
+								message.User,
+								_commitSource);
 			_currentRequests.Add(message.InternalCorrId, manager);
 			_currentTimedRequests.Add(message.InternalCorrId, Stopwatch.StartNew());
-			manager.Handle(message);
+			manager.Start();
 		}
 
 		public void Handle(ClientMessage.TransactionWrite message) {
-			var manager = new SingleAckRequestManager(_bus, _prepareTimeout, _betterOrdering);
+			var manager = new TransactionWrite(
+								_bus,
+								_prepareTimeout,
+								message.Envelope,
+								message.InternalCorrId,
+								message.CorrelationId,
+								message.Events,
+								message.TransactionId,
+								_commitSource);
 			_currentRequests.Add(message.InternalCorrId, manager);
 			_currentTimedRequests.Add(message.InternalCorrId, Stopwatch.StartNew());
-			manager.Handle(message);
+			manager.Start();
 		}
 
 		public void Handle(ClientMessage.TransactionCommit message) {
-			var manager = new TransactionCommitTwoPhaseRequestManager(_bus, _prepareCount, _prepareTimeout,
-				_commitTimeout, _betterOrdering);
+			var manager = new TransactionCommit(
+								_bus,
+								_prepareTimeout,
+								_commitTimeout,
+								message.Envelope,
+								message.InternalCorrId,
+								message.CorrelationId,
+								message.TransactionId,
+								_betterOrdering,
+								message.User,
+								_commitSource);
 			_currentRequests.Add(message.InternalCorrId, manager);
 			_currentTimedRequests.Add(message.InternalCorrId, Stopwatch.StartNew());
-			manager.Handle(message);
-		}
-
-		public void Handle(StorageMessage.RequestCompleted message) {
-			Stopwatch watch = null;
-			if (_currentTimedRequests.TryGetValue(message.CorrelationId, out watch)) {
-				HistogramService.SetValue(_requestManagerHistogram,
-					(long)((((double)watch.ElapsedTicks) / Stopwatch.Frequency) * 1000000000));
-				_currentTimedRequests.Remove(message.CorrelationId);
-			}
-			
-			if (!_currentRequests.Remove(message.CorrelationId))
-				throw new InvalidOperationException("Should never complete request twice.");
-		}
-
-		public void Handle(StorageMessage.CheckStreamAccessCompleted message) {
-			DispatchInternal(message.CorrelationId, message);
-		}
-
-		public void Handle(StorageMessage.AlreadyCommitted message) {
-			DispatchInternal(message.CorrelationId, message);
-		}
-
-		public void Handle(StorageMessage.PrepareAck message) {
-			DispatchInternal(message.CorrelationId, message);
-		}
-
-		public void Handle(StorageMessage.CommitReplicated message) {
-			DispatchInternal(message.CorrelationId, message);
+			manager.Start();
 		}
-
-		public void Handle(StorageMessage.WrongExpectedVersion message) {
-			DispatchInternal(message.CorrelationId, message);
-		}
-
-		public void Handle(StorageMessage.InvalidTransaction message) {
-			DispatchInternal(message.CorrelationId, message);
+		
+		
+		public void Handle(SystemMessage.StateChangeMessage message) {
+			//todo-clc: if we have become resigning master should all requests be actively disposed?
```

Comment:
```suggestion
			//TODO(clc): if we have become resigning master should all requests be actively disposed?
```

Yes we probably should.